src/backends/mysql/session.cpp

```diff
@@ -148,155 +47,62 @@ bool valid_int(const string & s)
     return true;
 }
 
-void parse_connect_string(const string & connectString,
-    string *host, bool *host_p,
-    string *user, bool *user_p,
-    string *password, bool *password_p,
-    string *db, bool *db_p,
-    string *unix_socket, bool *unix_socket_p,
-    int *port, bool *port_p, string *ssl_ca, bool *ssl_ca_p,
-    string *ssl_cert, bool *ssl_cert_p, string *ssl_key, bool *ssl_key_p,
-    int *local_infile, bool *local_infile_p,
-    string *charset, bool *charset_p)
-{
-    *host_p = false;
-    *user_p = false;
-    *password_p = false;
-    *db_p = false;
-    *unix_socket_p = false;
-    *port_p = false;
-    *ssl_ca_p = false;
-    *ssl_cert_p = false;
-    *ssl_key_p = false;
-    *local_infile_p = false;
-    *charset_p = false;
-    string err = "Malformed connection string.";
-    string::const_iterator i = connectString.begin(),
-        end = connectString.end();
-    while (i != end)
+} // namespace anonymous
+
+
+mysql_session_backend::mysql_session_backend(
+    connection_parameters const & params)
+{
+    string host, user, password, db, unix_socket, portstr, ssl_ca, ssl_cert, ssl_key,
+        charset, local_infstr;
+    int port = 0, local_infile = 0;
+
+    bool host_p = params.get_option("host", host);
+    bool user_p = params.get_option("user", user);
+    bool password_p = (params.get_option("pass", password) ||
+            params.get_option("password", password));
+    bool db_p = ( params.get_option("db", db) ||
+            params.get_option("dbname", db) ||
+            params.get_option("service", db) );
+    bool unix_socket_p = params.get_option("unix_socket", unix_socket);
+    bool port_p = params.get_option("port", portstr);
+    if (port_p)
     {
-        skip_white(&i, end, true);
-        if (i == end)
+        if (valid_int(portstr))
         {
-            return;
-        }
-        string par = param_name(&i, end);
-        skip_white(&i, end, false);
-        if (*i == '=')
-        {
-            ++i;
-        }
-        else
-        {
-            throw soci_error(err);
-        }
-        skip_white(&i, end, false);
-        string val = param_value(&i, end);
-        if (par == "port" and not *port_p)
-        {
-            if (not valid_int(val))
-            {
-                throw soci_error(err);
-            }
-            *port = std::atoi(val.c_str());
-            if (*port < 0)
+            port = std::atoi(portstr.c_str());
+            if (port < 0)
```

Comment:
Sorry, I didn't notice this didn't change. It's still wrong though, `atoi()` returns 0 if you pass it `"foobar"`, so we should test for it ideally.